### ISSUE TITLE
Offsetting multiple lines with holes does not work

### ISSUE THREAD

**Issue by Nevsden** [lelongg/geo-clipper#7]
Offsetting multiple polygons at once using the `MultiPolygon` struct does not work for me. I try to offset a shape with holes (see image).
In this shape, there is one outer polygon and four inner. I tried to offset multiple, different configurations of these five lines. 

Using all polygon | leaving out any configuration of the inner polygons:
![image](https://user-images.githubusercontent.com/23317083/105405967-107ee480-5c2c-11eb-83ce-450e5056f040.png)
![image](https://user-images.githubusercontent.com/23317083/105406052-30aea380-5c2c-11eb-8636-db9a18ee5229.png)

Leaving out the outer polygon:
![image](https://user-images.githubusercontent.com/23317083/105405901-f80eca00-5c2b-11eb-9a32-9cf20314ef97.png)

It seems as if the clipper library differs between outer and inner polygons. However I am not sure, if I acknowledge for this in my code. This observation matches my knowledge about the original AngusJ clipper library in Delphi. There one also had to accommodate for inner polygons.

Here is my code:
First I convert my own type of polygon into the `Polygon` struct. These are converted into the `Multipolygon` struct thereafter. I made sure, that it works. At last I call the trait function `offset` on the multi-polygon. The results can be derived from the images.

Here is my code 
``` rust
let polygons: Vec<Polygon<f64>> = lines
        .iter()
        .map(|line| Polygon::<f64>::from(line))
        .collect();
let multipolygon = MultiPolygon::<f64>::from(polygons);

// Compute offset   
let collection = multipolygon.offset(
    step as f64,
    JoinType::Square,
    EndType::ClosedPolygon,
    1E+06,
);
```

![image](https://user-images.githubusercontent.com/23317083/105405554-8fbfe880-5c2b-11eb-9119-ba899ce065f7.png)

**Comment by Nevsden** [comment]
In the documentation of the [clipper library](http://www.angusj.com/delphi/clipper/documentation/Docs/Units/ClipperLib/Classes/ClipperOffset/_Body.htm) it says

``` text
Preconditions for offsetting:
1. The orientations of closed paths must be consistent such that outer polygons share the same orientation, and any holes have the opposite orientation (ie non-zero filling). Open paths must be oriented with closed outer polygons.
2. Polygons must not self-intersect. 
```

Following the documentation, I reversed the point list for holes, but this had no effect on the upper offsetting results.

**Comment by Nevsden** [comment]
I checked into the documentation of `geo_types::Polygon` again and learned, that polygons contain a single exterior as well as multiple interiors of type `LineString`. So I gave it a try and wrote a wrapper, where I put all my lines into a single polygon object, filling its exterior and interiors variables.

Notice in the code, that I explicitely rearrange the orientation of interior and exterior lines to match the documentation of the angus library. sadly, it still does not work.

```rust
impl From<&Area> for Polygon<f64> {
    fn from(area: &Area) -> Self {
        // Compute exterior line
        let exterior = area.get_exterior();
        let exterior = LineString::from(exterior);
        // Compute set of interior lines aka holes
        let interiors = area.get_interiors();
        let interiors = interiors
            .into_iter()
            .map(|line| LineString::from(line))
            .collect();
        let polygon = Polygon::new(exterior, interiors);
        polygon.orient(Direction::Default)
    }
}
```

Sadly, this also did not work.

**Comment by Nevsden** [comment]
Update 25.01.2021:
### Debugging session: Pascal vs. C++
As I maintain the source code for a working wrapper around the AngusJ pascal library, also written in Pascal, I began a thorough debugging session checking both the working Pascal version and the C++ version with the clipper-sys wrapper around. Intermediate findings lead to a problem with the orientation of input lines. I tried to isolate this problem by correctly orientating all my lines according to the requirements of the offset function and commenting out source code, that would reorient any line. 

### My current guess is that a correct orientation of lines may do the trick
For the correct orientation of all lines I used the code, that I described in the [upper comment](https://github.com/lelongg/geo-clipper/issues/7#issuecomment-765341399). Then I commented out [this line](https://github.com/lelongg/clipper-sys/blob/61d5bcdc259185f657871e96a28eed8668509cc9/clipper/clipper.cpp#L3922) in the source code, which should reorient any path, which is a hole. At last, I checked both my working Pascal version and the C++ version against each other using vs code for debugging of the c++ code and Delphi Embacadero for the pascal code. 

### Narrowing down the problem
I found [this line](https://github.com/lelongg/clipper-sys/blob/61d5bcdc259185f657871e96a28eed8668509cc9/clipper/clipper.cpp#L4152) to be a good measure for proper orientation of the input paths. `if (m_sinA * m_delta < 0)`  works as follows: If the result is smaller zero, the corresponding line is shrunk, else the line is extended. Applying this rule to my [input shape](https://github.com/lelongg/geo-clipper/issues/7#issue-791456598) containing five lines, one outer line and four inner lines, the result of the upper source code line `m_sinA * m_delta` should be smaller zero for the outer line and bigger zero for any inner lines. Still, for some inner lines, it is smaller zero. This suggests, that either the orientation of the five lines is not correct, or there may be an error inside the `OffsetPoint` function or prior to the function.

**Comment by lelongg** [comment]
Thanks for your investigation, I have no time right now to do it myself but I will gladly approve a PR

**Comment by Nevsden** [comment]
Update 27.01.2021: @lelongg 
TLDR; The offset operator works. Still I would suggest to add some information to the crate documentation on how to integrate the traits correctly. 

While debugging the offset operation from this crate and after the offset operation has been finished I stumpled upon the resulting object of type `Polygon` having four lines within its inner variable `interior`. It seems that I am just too lazy to read the documentation and thus have not been able to correctly implement the trait from this crate.
